polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+    resolve_order: Vec<String>,
+    result_id: String,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+pub struct FilterPlan {
+    result_sets: Vec<ResultSet>,
+}
+
+impl FilterPlan {
+    pub fn explain(&self) {
+        eprintln!("UNION");
+        for (i, result_set) in self.result_sets.iter().enumerate() {
+            eprintln!("  =Result Set: {}=", i);
+            for id in &result_set.resolve_order {
+                let fetch_request = result_set.requests.get(id).unwrap();
+                eprintln!("    {}: Fetch {}", id, fetch_request.class_tag);
+                for constraint in &fetch_request.constraints {
+                    let op = match constraint.kind {
+                        ConstraintKind::Eq => "=".to_owned(),
+                        ConstraintKind::In => "in".to_owned(),
+                        ConstraintKind::Contains => "contains".to_owned(),
+                    };
+                    let field = constraint.field.clone();
+                    let value = match &constraint.value {
+                        ConstraintValue::Term(t) => t.to_polar(),
+                        ConstraintValue::Field(f) => format!("FIELD({})", f),
+                        ConstraintValue::Ref(r) => {
+                            let mut s = "REF(".to_owned();
+                            if let Some(field) = &r.field {
+                                s.push_str(&format!("field {} of ", field));
+                            }
+                            s.push_str(&format!("result {})", r.result_id));
+                            s
+                        }
+                    };
+                    eprintln!("          {} {} {}", field, op, value);
+                }
+            }
+        }
+    }
+}
+
+pub type Types = HashMap<String, HashMap<String, Type>>;
+pub type PartialResults = Vec<ResultEvent>;
+
+#[derive(Debug)]
+struct VarInfo {
+    cycles: Vec<(Symbol, Symbol)>,                      // x = y
+    types: Vec<(Symbol, String)>,                       // x matches XClass
+    eq_values: Vec<(Symbol, Term)>,                     // x = 1;
+    contained_values: Vec<(Term, Symbol)>,              // 1 in x
+    field_relationships: Vec<(Symbol, String, Symbol)>, // x.a = y
+    in_relationships: Vec<(Symbol, Symbol)>,            // x in y
+}
+
+// @TODO(steve): Better way to handle these checks than just unwraps and asserts.
+
+fn process_result(exp: &Operation) -> VarInfo {
+    let mut var_info = VarInfo {
+        cycles: vec![],
+        types: vec![],
+        eq_values: vec![],
+        contained_values: vec![],
+        field_relationships: vec![],
+        in_relationships: vec![],
+    };
+    process_exp(&mut var_info, exp);
+    var_info
+}
+
+fn dot_var(var_info: &mut VarInfo, var: Term, field: &Term) -> Symbol {
+    let mut var = var;
+    while let Ok(Operation {
+        operator: Operator::Dot,
+        args,
+    }) = var.value().as_expression()
```

Comment:
Do we need this loop here & also the recursive call in `process_exp`? I didn't think about this too long but it seems like this might run several times for every part of the expression.